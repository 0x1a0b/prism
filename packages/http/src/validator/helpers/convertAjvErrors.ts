import { ValidationSeverity } from '@stoplight/prism-core/types';
import { ErrorObject } from 'ajv';

export function convertAjvErrors(
  errors: ErrorObject[]|undefined|null,
  pathPrefix: string,
  severity: ValidationSeverity
) {
  if (!errors) {
    return [];
  }

  return errors.map(error => ({
    path: [pathPrefix, ...error.dataPath.split('.').slice(1)],
    name: error.keyword || '',
    summary: error.message || '',
    message: error.message || '',
    severity,
  }));
}
