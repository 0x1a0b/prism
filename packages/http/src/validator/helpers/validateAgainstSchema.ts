import { IValidation } from '@stoplight/prism-core';
import { ValidationSeverity } from '@stoplight/prism-core/types';
import { ISchema } from '@stoplight/types/schema';
import * as Ajv from 'ajv';

import { convertAjvErrors } from './convertAjvErrors';

const ajv = new Ajv({ allErrors: true, messages: true });

export function validateAgainstSchema(value: any, schema: ISchema, prefix: string): IValidation[] {
  const validate = ajv.compile(schema);

  if (!validate(value)) {
    return convertAjvErrors(validate.errors, ValidationSeverity.ERROR).map(error =>
      Object.assign({}, error, { path: [prefix, ...error.path] })
    );
  }

  return [];
}
