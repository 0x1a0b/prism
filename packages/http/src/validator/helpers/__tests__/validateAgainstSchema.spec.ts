import * as convertAjvErrorsModule from '../convertAjvErrors';
import { validateAgainstSchema } from '../validateAgainstSchema';

describe('validateAgainstSchema()', () => {
  beforeEach(() => {
    jest.clearAllMocks();
    jest.spyOn(convertAjvErrorsModule, 'convertAjvErrors');
  });

  describe('has no validation errors', () => {
    it('returns no validation errors', () => {
      expect(validateAgainstSchema('test', { type: 'string' }, 'pfx')).toEqual([]);
      expect(convertAjvErrorsModule.convertAjvErrors).not.toHaveBeenCalled();
    });
  });

  describe('has validation errors', () => {
    it('returns validation errors', () => {
      jest.spyOn(convertAjvErrorsModule, 'convertAjvErrors').mockImplementationOnce(() => [
        {
          message: 'should be number',
          name: 'type',
          path: [],
          severity: 'error',
          summary: 'should be number',
        },
      ]);
      expect(validateAgainstSchema('test', { type: 'number' }, 'pfx')).toMatchSnapshot();
      expect(convertAjvErrorsModule.convertAjvErrors).toHaveBeenCalledWith(
        [
          {
            dataPath: '',
            keyword: 'type',
            message: 'should be number',
            params: { type: 'number' },
            schemaPath: '#/type',
          },
        ],
        'error'
      );
    });
  });
});
