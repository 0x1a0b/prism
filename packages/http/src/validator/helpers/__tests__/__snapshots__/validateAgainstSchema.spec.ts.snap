// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`validateAgainstSchema() has validation errors returns validation errors 1`] = `
Array [
  Object {
    "message": "should be number",
    "name": "type",
    "path": Array [
      "pfx",
    ],
    "severity": "error",
    "summary": "should be number",
  },
]
`;
